src/mlpack/methods/ann/rbm/rbm_impl.hpp

```diff
@@ -0,0 +1,229 @@
+#include "rbm.hpp"
+
+#include <mlpack/core.hpp>
+
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial neural networks. */ {
+
+template<typename InitializationRuleType, typename RBMPolicy>
+RBM<InitializationRuleType, RBMPolicy>::RBM(
+    arma::Mat<ElemType> predictors,
+    InitializationRuleType initializeRule,
+    RBMPolicy rbmPolicy,
+    const size_t numSteps,
+    const size_t negSteps,
+    const bool useMonitoringCost,
+    const bool persistence):
+    predictors(std::move(predictors)),
+    initializeRule(initializeRule),
+    rbmPolicy(rbmPolicy),
+    numSteps(numSteps),
+    negSteps(negSteps),
+    useMonitoringCost(useMonitoringCost),
+    persistence(persistence),
+    reset(false)
+{
+  numFunctions = this->predictors.n_cols;
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+void RBM<InitializationRuleType, RBMPolicy>::Reset()
+{
+  size_t weight = rbmPolicy.Parameters().n_elem;
+
+  positiveGradient.set_size(weight, 1);
+  negativeGradient.set_size(weight, 1);
+  negativeSamples.set_size(rbmPolicy.VisibleSize(), 1);
+  tempNegativeGradient.set_size(weight, 1);
+
+  parameter.zeros(weight, 1);
+  positiveGradient.zeros();
+  negativeGradient.zeros();
+  tempNegativeGradient.zeros();
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
+
+  rbmPolicy.Parameters() = arma::Mat<ElemType>(parameter.memptr(), weight, 1,
+      false, false);
+  rbmPolicy.Reset();
+
+  reset = true;
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+template<typename OptimizerType>
+void RBM<InitializationRuleType, RBMPolicy>::Train(
+    OptimizerType& optimizer)
+{
+  if (!reset)
+    Reset();
+  optimizer.Optimize(*this, parameter);
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+double RBM<InitializationRuleType, RBMPolicy>::
+    FreeEnergy(arma::Mat<ElemType>&& input)
+{
+  return rbmPolicy.FreeEnergy(std::move(input));
+}
+
+template<typename InitializationRuleType, typename RBMPolicy>
+double RBM<InitializationRuleType, RBMPolicy>::Evaluate(
+    const arma::Mat<ElemType>& /* parameters*/,
+    const size_t i,
+    const size_t batchSize)
+{
+  if (!useMonitoringCost)
+  {
+    Gibbs(std::move(predictors.col(i)), std::move(negativeSamples));
+    return std::fabs(FreeEnergy(std::move(predictors.col(i))) -
+        FreeEnergy(std::move(negativeSamples)));
+  }
+  else
+  {
+    if (persistence)
+    {
+      return rbmPolicy.Evaluate(predictors, i, batchSize);
+    }
+    else
+    {
+      // Mean Squared Error
+      rbmPolicy.SampleHidden(std::move(predictors.col(i)),
+          std::move(hiddenReconstruction));
+      rbmPolicy.SampleVisible(std::move(hiddenReconstruction),
+          std::move(visibleReconstruction));
+      return arma::accu(arma::pow(visibleReconstruction -
+                                  predictors.col(i),
+          2)) / hiddenReconstruction.n_rows;
+    }
```

Comment:
@lozhnikov Could you comment why are we using mean squared error here? I don't see it being mentioned anywhere in the deepearning.net tutorial?

Reply:
@ShikharJ Actually, I think you can get rid of
```
if (persistence)
{
  return rbmPolicy.Evaluate(predictors, i, batchSize);
}
```
either.